### PR TITLE
DD-254 - Refactor dd-dans-deposit-to-dataverse to use dans-dataverse-scala-lib

### PR DESCRIPTION
**Pull request by ChrisMuller80**
Fixes [DD-254](https://drivenbydata.atlassian.net/browse/DD-254)

# Description of changes
Refactoring for the purpose of using `dans-dataverse-scala-lib`

# How to test
run` run-service.sh` (with a valid deposit in the watched inbox)

# Related PRs 
(Add links)
* 

# Notify
@DANS-KNAW/dataversedans

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala [hunk 1]
```diff
@@ -51,28 +48,29 @@ case class DepositIngestTask(deposit: Deposit, dansBagValidator: DansBagValidato
     logger.info(s"Ingesting $deposit into Dataverse")
 
     for {
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:50]: Oops?

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala [hunk 2]
```diff
@@ -82,29 +80,26 @@ case class DepositIngestTask(deposit: Deposit, dansBagValidator: DansBagValidato
     // TODO: delete draft if something went wrong
   }
 
-  private def formatViolation(v: (String, String)): String = v match {
-    case (nr, msg) => s" - [$nr] $msg"
+  private def getPersistentId(response: DataverseResponse[DatasetCreationResult]): Try[String] = Try {
+    response.data.get.persistentId.toString
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:84]: response.data gives you a Try and you need a Try in the return value of getPersistentId; then why call `get`? You can just do

```
    response.data.map(_.persistentId)
```

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala [hunk 3]
```diff
@@ -82,29 +80,26 @@ case class DepositIngestTask(deposit: Deposit, dansBagValidator: DansBagValidato
...
+    (JsonMethods.parse(json) \ "datasetVersion").extract[DatasetVersion]
   }
 
   private def uploadFilesToDataset(datasetId: String): Try[Unit] = {
     trace(datasetId)
     for {
       filesXml <- deposit.tryFilesXml
       ddm <- deposit.tryDdm
-      defaultRestrict <- Try { (ddm \ "profile" \ "accessRights").headOption.forall(AccessRights toDefaultRestrict) }
+      defaultRestrict = (ddm \ "profile" \ "accessRights").headOption.forall(AccessRights toDefaultRestrict)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:96]: Why this change?

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala [hunk 4]
```diff
@@ -51,28 +48,29 @@ case class DepositIngestTask(deposit: Deposit, dansBagValidator: DansBagValidato
...
+      //        if (!validationResult.isCompliant) throw RejectedDepositException(deposit,
+      //          s"""
+      //             |Bag was not valid according to Profile Version ${ validationResult.profileVersion }.
+      //             |Violations:
+      //             |${ validationResult.ruleViolations.map(_.map(formatViolation).mkString("\n")).getOrElse("") }
+      //          """.stripMargin)
+      //      }
       ddm <- deposit.tryDdm
       dataverseDataset <- mapper.toDataverseDataset(ddm, deposit.vaultMetadata)
       json = Serialization.writePretty(dataverseDataset)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:62]: Why first serialize and then extract again?

File: src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala [hunk 5]
```diff
@@ -39,60 +37,34 @@ package object json {
...
-  def createPrimitiveFieldMultipleValues(name: String, values: List[String]): PrimitiveFieldMultipleValues = {
-    PrimitiveFieldMultipleValues(name, multiple = true, "primitive", values)
+  def createPrimitiveFieldMultipleValues(name: String, values: List[String]): PrimitiveMultipleValueField = {
+    PrimitiveMultipleValueField(name, values)
   }
 
-  def createCvFieldSingleValue(name: String, value: String): PrimitiveFieldSingleValue = {
-    PrimitiveFieldSingleValue(name, multiple = false, "controlledVocabulary", value)
+  def createCvFieldSingleValue(name: String, value: String): ControlledSingleValueField = {
+    ControlledSingleValueField("controlledVocabulary", name, multiple = false, value)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala:56]: I see that I forgot to change this constructor in the library. Wil do that.

File: src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala [hunk 6]
```diff
@@ -39,60 +37,34 @@ package object json {
       fields.put(name, createCvFieldSingleValue(name, value))
     }
 
-    def addCompoundField(name: String, value: Map[String, Field]): Unit = {
+    def addCompoundField(name: String, value: Map[String, MetadataField]): Unit = {
       fields.put(name, createCompoundFieldSingleValue(name, value))
     }
 
     def toJsonObject: JsonObject = fields.toMap
   }
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala:46]: The helper functions below have no added value anymore.

File: src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala [hunk 7]
```diff
@@ -15,21 +15,19 @@
  */
 package nl.knaw.dans.easy.dd2d.dataverse
 
-import better.files.File
+import nl.knaw.dans.lib.dataverse.model.dataset.{ CompoundField, ControlledMultipleValueField, ControlledSingleValueField, MetadataField, PrimitiveMultipleValueField, PrimitiveSingleValueField }
 import org.json4s.{ DefaultFormats, Formats }
 
 import scala.collection.mutable
 
 package object json {
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.dd2d/dataverse/json/package.scala:23]: Maybe we coudl get rid of this package altogether. The FieldMap is still handy though. In any case, we should move it. There is now an empty parent package "dataverse".  FieldMap is exclusively used from the mapping package, so how about moving it to the package object of that package?

**Review by janvanmansum — CHANGES_REQUESTED** [review]
Not bad at all! Still, there is room for improvement (and one blooper, you left the validation code commented out).

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala (hunk 1 above)
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:50]: Oops

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala (hunk 3 above)
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:96]: Joke told me that this is a more concise way of doing the same. I may have misunderstood.

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala (hunk 4 above)
Comment (ChrisMuller80) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:62]: Fixed. I removed the debug statement as well (since there is no need to serialize). Do we need to see the Dataset json?

File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala (hunk 3 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.dd2d/DepositIngestTask.scala:90]: No, it seems you (and @jo-pol) are right. I was not sure if an exception raised on the right-hand side of the `=` would be handled the same as in a `<- Try { ...}` so I did a little test in a scala workbook:
```scala
import scala.util.Try

// With  <- Try {}
for {
  _ <- Try { throw new RuntimeException("test")}
} yield()


// With =
try
{
  for {
    _ <- Try { } // Have to put this here, otherwise it does not compile
    _ = throw new RuntimeException("test")
  } yield ()
}
catch {
  case e: Exception => println("caught")
}

```
Both return a `Failure` and no "caught" is printed, so that proves your point.

**Review by janvanmansum — APPROVED** [review]
